README.md
NCU-CLUB
========


CHANGELOG.md
Version 1.0
========
Version 1.2

Fix issues
- clickable link in list item
- decrease duration of animation in drawer

--------
Version 1.3

Fix issues
- link of attachment
- null club name
- redundant time format

--------
Version 1.4

Fix issues
- bug of adding new calendar event

--------
Version 1.5

Fix bugs

--------
Version 1.6

Change icon and app name

--------
Version 1.7

update api client
--------

Version 1.8

fix null problem of start column 
--------

Version 1.9

Update to SDK 23
Using HTTPUrlConnection instead of apache httpClient

app/src/main/java/evan/ncu/club/BitmapCache.java
/*
 * Copyright 2014 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package evan.ncu.club;

import android.graphics.Bitmap;
import android.util.LruCache;

public class BitmapCache extends LruCache<Integer, Bitmap> {

    private static final int KILO = 1024;
    private static final int MEMORY_FACTOR = 2 * KILO;

    public BitmapCache() {
        super((int) (Runtime.getRuntime().maxMemory() / MEMORY_FACTOR));
    }

    @Override
    protected int sizeOf(final Integer key, final Bitmap value) {
        return value.getRowBytes() * value.getHeight() / KILO;
    }
}


app/src/main/java/evan/ncu/club/GeneralBoard.java
package evan.ncu.club;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.AlertDialog.Builder;
import android.app.Fragment;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnCancelListener;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Base64;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.Toast;

import com.nhaarman.listviewanimations.appearance.simple.AlphaInAnimationAdapter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class GeneralBoard extends Fragment {

    private static final int INITIAL_DELAY_MILLIS = 500;
    private ListView m_ListView;
    public MyExpandableListItemAdapter listAdapter;
    public AlphaInAnimationAdapter alphaInAnimationAdapter;
    public ArrayList<ListData> itemsArray;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_latest, container,
                false);
        m_ListView = (ListView) rootView.findViewById(R.id.m_ListView);

        new RetrieveList().execute();
        return rootView;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        ((MainActivity) activity).onSectionAttached(1);
    }

    private class RetrieveList extends
            AsyncTask<Void, Void, Void> {
        private ProgressDialog progressDialog;
        final ConnectivityManager conMgr = (ConnectivityManager) getActivity()
                .getApplicationContext().getSystemService(
                        Context.CONNECTIVITY_SERVICE);
        final NetworkInfo activeNetwork = conMgr.getActiveNetworkInfo();
        private boolean isNetwork = true;
        private boolean serverStatus = true;
        @Override
        protected void onPreExecute() {
            progressDialog = new ProgressDialog(getActivity());
            progressDialog.setMessage("讀取中");
            progressDialog.setCancelable(false);
            if (progressDialog != null)
                progressDialog.show();
            progressDialog.setOnCancelListener(new OnCancelListener() {
                public void onCancel(DialogInterface arg0) {
                }
            });
            itemsArray = new ArrayList<ListData>();
        }

        @Override
        protected Void doInBackground(Void... arg0) {

            if (activeNetwork != null && activeNetwork.isConnected()) {
                String result = "";
                try {
                    String url = "https://api.cc.ncu.edu.tw/activity/v1/announces/?type=common&size=40";
                    URL myURL = new URL(url);
                    HttpURLConnection myURLConnection = (HttpURLConnection)myURL.openConnection();
                    myURLConnection.setRequestMethod("GET");
                    myURLConnection.setRequestProperty("X-NCU-API-TOKEN", getString(R.string.ncu_api_token));
                    myURLConnection.setUseCaches(false);
                    myURLConnection.setDoInput(true);
                    myURLConnection.connect();

                    int response = myURLConnection.getResponseCode();
                    BufferedReader r = new BufferedReader(new InputStreamReader(myURLConnection.getInputStream()));
                    StringBuilder total = new StringBuilder();
                    String line;
                    while ((line = r.readLine()) != null) {
                        total.append(line);
                    }
                    result = total.toString();

                    if (response == 200) {
                        Log.d("Debug", result);
                        JSONArray jsonArray = new JSONArray(result);

                        for (int i = 0; i < jsonArray.length(); i++) {
                            try {
                                JSONObject jsonData = jsonArray.getJSONObject(i);
                                String name = jsonData.getString("title");
                                String content = jsonData.getString("content");
                                String time = jsonData.getString("time");
                                String link = jsonData.getString("attachment");
                                String time_parsed;
                                Date date;
                                SimpleDateFormat simple = new java.text.SimpleDateFormat();
                                simple.applyPattern("yyyy-MM-dd HH:mm");
                                date = simple.parse(time);
                                simple.applyPattern("yyyy-MM-dd");
                                time_parsed = simple.format(date);
                                ListData listData;
                                if (link != "null")
                                    listData = new ListData(name, time, content + "\n" + link, time_parsed);
                                else
                                    listData = new ListData(name, time, content, time_parsed);
                                itemsArray.add(listData);
                            }
                            catch (JSONException je) {
                                Log.e("Debug", je.toString());
                            }
                        }
                    }


                } catch (Exception e) {
                    e.printStackTrace();
                    serverStatus = false;
                }

            } else {
                Log.e("Debug", "No network");
                isNetwork = false;
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void arg0) {
            if (!isNetwork) {
                Builder MyAlertDialog = new AlertDialog.Builder(getActivity());
                MyAlertDialog.setTitle("錯誤~");
                MyAlertDialog.setMessage("需要網際網路連線");
                MyAlertDialog.show();
            }
            if(!serverStatus){
                Toast.makeText(getActivity(), "無法與伺服器連線", Toast.LENGTH_SHORT).show();
            }
            listAdapter = new MyExpandableListItemAdapter(
                    getActivity(), itemsArray);
            alphaInAnimationAdapter = new AlphaInAnimationAdapter(listAdapter);
            alphaInAnimationAdapter.setAbsListView(m_ListView);
            assert alphaInAnimationAdapter.getViewAnimator() != null;
            alphaInAnimationAdapter.getViewAnimator().setInitialDelayMillis(INITIAL_DELAY_MILLIS);
            m_ListView.setAdapter(alphaInAnimationAdapter);

            if (progressDialog != null)
                progressDialog.dismiss();

        }
    }

}


app/src/main/java/evan/ncu/club/MainActivity.java
package evan.ncu.club;

import android.app.Fragment;
import android.app.FragmentManager;
import android.os.Bundle;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.nhaarman.listviewanimations.appearance.simple.SwingLeftInAnimationAdapter;

public class MainActivity extends ActionBarActivity{

    private DrawerLayout mDrawerLayout;
    private ActionBarDrawerToggle mDrawerToggle;
	private CharSequence mTitle;
    private ListView mDrawerListView;
    private int mCurrentSelectedPosition = 0;
    private SwingLeftInAnimationAdapter mAnimAdapter;
    private static final int INITIAL_DELAY_MILLIS = 500;
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		mTitle = getTitle();

		// Set up the drawer.
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.inflateMenu(R.menu.main);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
            mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer);
            // 實作 drawer toggle 並放入 toolbar
            mDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close){
                @Override
            public void onDrawerClosed(View drawerView) {
                mDrawerListView.setVisibility(View.GONE);
                super.onDrawerClosed(drawerView);
                restoreActionBar();
                invalidateOptionsMenu();
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                restoreActionBar();
                invalidateOptionsMenu();
            }
        };
        mDrawerToggle.syncState();
        mDrawerLayout.setDrawerListener(mDrawerToggle);
        mDrawerLayout.setDrawerShadow(R.drawable.drawer_shadow,
                GravityCompat.START);

        initDrawer();
        onNavigationDrawerItemSelected(0);
        getSupportActionBar().setTitle(getString(R.string.title_section1));
	}

	public void onNavigationDrawerItemSelected(int position) {
		// update the main content by replacing fragments
		Fragment fragment = new GeneralBoard();
		FragmentManager fragmentManager = getFragmentManager();
		
		 switch(position) {
	        case 0:
                fragment = new LatestEvent();
	            break;
	        case 1:
                fragment = new GeneralBoard();
	            break;
	        case 2:
	            fragment = new LatestSchedule();
	            break;
	        
	    }
	    fragmentManager.beginTransaction()
	        .replace(R.id.container, fragment)
	        .commit();
	}

	public void onSectionAttached(int number) {
		switch (number) {
		case 0:
			mTitle = getString(R.string.title_section1);
			break;
		case 1:
			mTitle = getString(R.string.title_section2);
			break;
		case 2:
			mTitle = getString(R.string.title_section3);
			break;
		}
	}

	public void restoreActionBar() {
		ActionBar actionBar = getSupportActionBar();
		actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_STANDARD);
		actionBar.setDisplayShowTitleEnabled(true);
		actionBar.setTitle(mTitle);
	}
    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        // If the nav drawer is open, hide action items related to the content view
        boolean drawerOpen = mDrawerLayout.isDrawerOpen(mDrawerListView);
        if(drawerOpen){

            mAnimAdapter = new SwingLeftInAnimationAdapter(new ArrayAdapter<String>(getApplicationContext(),
                    android.R.layout.simple_list_item_activated_1,
                    android.R.id.text1, new String[] {
                    getString(R.string.title_section1),
                    getString(R.string.title_section2),
                    getString(R.string.title_section3), }));

            mAnimAdapter.setAbsListView(mDrawerListView);
            assert mAnimAdapter.getViewAnimator() != null;
            mAnimAdapter.getViewAnimator().setInitialDelayMillis(0);
            mAnimAdapter.getViewAnimator().setAnimationDurationMillis(250);

            mDrawerListView.setAdapter(mAnimAdapter);
            if (mCurrentSelectedPosition >= 0) {
                mDrawerListView.setItemChecked(mCurrentSelectedPosition, true);
            }

        }
        else{
            mDrawerListView.setAdapter(null);
        }
        return super.onPrepareOptionsMenu(menu);
    }
	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
//		if (!mNavigationDrawerFragment.isDrawerOpen()) {
//			// Only show items in the action bar relevant to this screen
//			// if the drawer is not showing. Otherwise, let the drawer
//			// decide what to show in the action bar.
//			getMenuInflater().inflate(R.menu.main, menu);
//			restoreActionBar();
//			return true;
//		}
		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		// Handle action bar item clicks here. The action bar will
		// automatically handle clicks on the Home/Up button, so long
		// as you specify a parent activity in AndroidManifest.xml.
		int id = item.getItemId();
//		if (id == R.id.action_settings) {
//			return true;
//		}
		return super.onOptionsItemSelected(item);
	}

    public void initDrawer(){
        mDrawerListView = (ListView) findViewById(R.id.drawer_list);
        mDrawerListView
                .setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view,
                                            int position, long id) {
                        selectItem(position);
                    }
                });
    }
    private void selectItem(int position) {
        mCurrentSelectedPosition = position;
        if (mDrawerLayout != null) {
            mDrawerLayout.closeDrawer(mDrawerListView);
        }
        onNavigationDrawerItemSelected(position);
    }
}


app/src/main/java/evan/ncu/club/LatestSchedule.java
package evan.ncu.club;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Fragment;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.Toast;

import com.nhaarman.listviewanimations.appearance.simple.AlphaInAnimationAdapter;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class LatestSchedule extends Fragment {

    private static final int INITIAL_DELAY_MILLIS = 500;
    private ListView m_ListView;
    public MyExpandableListItemAdapter listAdapter;
    public AlphaInAnimationAdapter alphaInAnimationAdapter;
    public ArrayList<ListData> itemsArray;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_latest, container,
                false);
        m_ListView = (ListView) rootView.findViewById(R.id.m_ListView);

        new RetrieveList().execute();
        return rootView;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        ((MainActivity) activity).onSectionAttached(2);
    }

    private class RetrieveList extends
            AsyncTask<Void, Void, Void> {
        private ProgressDialog progressDialog;
        final ConnectivityManager conMgr = (ConnectivityManager) getActivity()
                .getApplicationContext().getSystemService(
                        Context.CONNECTIVITY_SERVICE);
        final NetworkInfo activeNetwork = conMgr.getActiveNetworkInfo();
        private boolean isNetwork = true;
        private boolean serverStatus = true;
        @Override
        protected void onPreExecute() {
            progressDialog = new ProgressDialog(getActivity());
            progressDialog.setMessage("讀取中");
            progressDialog.setCancelable(false);
            if (progressDialog != null)
                progressDialog.show();
            progressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
                public void onCancel(DialogInterface arg0) {
                }
            });
            itemsArray = new ArrayList<ListData>();
        }

        @Override
        protected Void doInBackground(Void... arg0) {

            if (activeNetwork != null && activeNetwork.isConnected()) {
                String result = "";
                try {

                    String url = "https://api.cc.ncu.edu.tw/activity/v1/announces?type=group&size=40";
                    URL myURL = new URL(url);
                    HttpURLConnection myURLConnection = (HttpURLConnection)myURL.openConnection();
                    myURLConnection.setRequestMethod("GET");
                    myURLConnection.setRequestProperty("X-NCU-API-TOKEN", getString(R.string.ncu_api_token));
                    myURLConnection.setUseCaches(false);
                    myURLConnection.setDoInput(true);
                    myURLConnection.connect();

                    int response = myURLConnection.getResponseCode();
                    BufferedReader r = new BufferedReader(new InputStreamReader(myURLConnection.getInputStream()));
                    StringBuilder total = new StringBuilder();
                    String line;
                    while ((line = r.readLine()) != null) {
                        total.append(line);
                    }
                    result = total.toString();

                    if (response == 200) {

                        JSONArray jsonArray = new JSONArray(result);

                        for (int i = 0; i < jsonArray.length(); i++) {
                            try {
                                JSONObject jsonData = jsonArray.getJSONObject(i);
                                String name = jsonData.getString("title");
                                String content = jsonData.getString("content");
                                String time = jsonData.getString("time");
                                String link = jsonData.getString("attachment");
                                String time_parsed;
                                Date date;
                                SimpleDateFormat simple = new java.text.SimpleDateFormat();
                                simple.applyPattern("yyyy-MM-dd HH:mm");
                                date = simple.parse(time);
                                simple.applyPattern("yyyy-MM-dd");
                                time_parsed = simple.format(date);
                                ListData listData;
                                if (link != "null")
                                    listData = new ListData(name, time, content + "\n" + link, time_parsed);
                                else
                                    listData = new ListData(name, time, content, time_parsed);
                                itemsArray.add(listData);
                            }
                            catch (JSONException je) {
                                Log.e("Debug", je.toString());
                            }
                        }
                    }


                } catch (Exception e) {
                    e.printStackTrace();
                    serverStatus = false;
                }
            } else {
                Log.e("Debug", "No network");
                isNetwork = false;
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void arg0) {
            if (!isNetwork) {
                AlertDialog.Builder MyAlertDialog = new AlertDialog.Builder(getActivity());
                MyAlertDialog.setTitle("錯誤~");
                MyAlertDialog.setMessage("需要網際網路連線");
                MyAlertDialog.show();
            }
            if(!serverStatus){
                Toast.makeText(getActivity(), "無法與伺服器連線", Toast.LENGTH_SHORT).show();
            }
            listAdapter = new MyExpandableListItemAdapter(
                    getActivity(), itemsArray);
            alphaInAnimationAdapter = new AlphaInAnimationAdapter(listAdapter);
            alphaInAnimationAdapter.setAbsListView(m_ListView);
            assert alphaInAnimationAdapter.getViewAnimator() != null;
            alphaInAnimationAdapter.getViewAnimator().setInitialDelayMillis(INITIAL_DELAY_MILLIS);
            m_ListView.setAdapter(alphaInAnimationAdapter);

            if (progressDialog != null)
                progressDialog.dismiss();

        }
    }

}


app/src/main/java/evan/ncu/club/MyExpandableListItemAdapter.java
package evan.ncu.club;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

//import com.nhaarman.listviewanimations.itemmanipulation.expandablelistitem.ExpandableListItemAdapter;

/**
 * Created by Evany on 2014/11/2.
 */
public class MyExpandableListItemAdapter extends
        ExpandableListItemAdapter<ListData> {
    private final Context mContext;
    private final BitmapCache mMemoryCache;
    /**
     * Creates a new ExpandableListItemAdapter with the specified list, or
     * an empty list if items == null.
     */
    public MyExpandableListItemAdapter(final Context context,
                                        final ArrayList<ListData> items) {

        super(context, R.layout.expandablelistitem_card,
                R.id.expandablelistitem_card_title,
                R.id.expandablelistitem_card_content, items);
        mContext = context;
        mMemoryCache = new BitmapCache();
        //setActionViewResId(R.id.expandablelistitem_card_title);

    }

    @Override
    public View getTitleView(final int position, final View convertView,
                             final ViewGroup parent) {
        View v = LayoutInflater.from(mContext).inflate(
                R.layout.expandablelistitem_title, null);
        TextView tv = (TextView) v.findViewById(R.id.item_Title);
        if (tv == null) {
            tv = new TextView(mContext);
        }
        tv.setText(getItem(position).title);
        tv.setTextColor(Color.BLACK);
        tv.setAlpha(0.87f);
        tv.setTextSize(16);

        tv = (TextView) v.findViewById(R.id.item_Date);
        tv.setText(getItem(position).time_parsed);
        tv.setTextColor(Color.BLACK);
        tv.setAlpha(0.54f);
        tv.setTextSize(14);
        Button addButton = (Button) v.findViewById(R.id.Addbutton);
        addButton.setOnClickListener(new Button.OnClickListener(){

            @Override
            public void onClick(View v) {

                try {
                    addCalendarEvent(getItem(position).title, getItem(position).content, getItem(position).time);
                } catch (ParseException e) {
                    Log.e("Debug",e.toString());
                    e.printStackTrace();
                }

            }

        });

        return v;
    }

    @Override
    public View getContentView(final int position, final View convertView,
                               final ViewGroup parent) {

        View v = LayoutInflater.from(mContext).inflate(
                R.layout.expandablelistitem_content, parent, false);

        TextView textView = (TextView) v.findViewById(R.id.contentTextView);
        textView.setText(getItem(position).content);
        textView.setTextColor(Color.BLACK);
        textView.setAlpha(0.53f);

        return v;
    }

    private void addBitmapToMemoryCache(final int key, final Bitmap bitmap) {
        if (getBitmapFromMemCache(key) == null) {
            mMemoryCache.put(key, bitmap);
        }
    }

    private Bitmap getBitmapFromMemCache(final int key) {
        return mMemoryCache.get(key);
    }

    public void addCalendarEvent(String title, String description, String time) throws ParseException {

        Calendar cal = Calendar.getInstance();
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd hh:mm");

        Date date = df.parse(time);
        cal.setTime(date);
        Intent intent = new Intent(Intent.ACTION_EDIT);
        intent.setType("vnd.android.cursor.item/event");
        intent.putExtra("beginTime", cal.getTimeInMillis());
        intent.putExtra("allDay", true);
        intent.putExtra("endTime", cal.getTimeInMillis()+60*60*1000);
        intent.putExtra("title", title);
        intent.putExtra("description", description);
        mContext.startActivity(intent);
    }


}


app/src/main/java/evan/ncu/club/LatestEvent.java
package evan.ncu.club;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Fragment;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.Toast;

import com.nhaarman.listviewanimations.appearance.simple.AlphaInAnimationAdapter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class LatestEvent extends Fragment {
    private static final int INITIAL_DELAY_MILLIS = 500;
    private ListView m_ListView;
    public MyExpandableListItemAdapter listAdapter;
    public AlphaInAnimationAdapter alphaInAnimationAdapter;
    public ArrayList<ListData> itemsArray;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_latest, container,
                false);
        m_ListView = (ListView) rootView.findViewById(R.id.m_ListView);

        new RetrieveList().execute();
        return rootView;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        ((MainActivity) activity).onSectionAttached(0);
    }

    private class RetrieveList extends
            AsyncTask<Void, Void, Void> {
        private ProgressDialog progressDialog;
        final ConnectivityManager conMgr = (ConnectivityManager) getActivity()
                .getApplicationContext().getSystemService(
                        Context.CONNECTIVITY_SERVICE);
        final NetworkInfo activeNetwork = conMgr.getActiveNetworkInfo();
        private boolean isNetwork = true;
        private boolean serverStatus = true;
        @Override
        protected void onPreExecute() {
            progressDialog = new ProgressDialog(getActivity());
            progressDialog.setMessage("讀取中");
            progressDialog.setCancelable(false);
            if (progressDialog != null)
                progressDialog.show();
            progressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
                public void onCancel(DialogInterface arg0) {
                }
            });
            itemsArray = new ArrayList<ListData>();
        }

        @Override
        protected Void doInBackground(Void... arg0) {
            if (activeNetwork != null && activeNetwork.isConnected()) {
                String result = "";
                try {

                    String url = "https://api.cc.ncu.edu.tw/activity/v1/activities?size=40";
                    URL myURL = new URL(url);
                    HttpURLConnection myURLConnection = (HttpURLConnection)myURL.openConnection();
                    myURLConnection.setRequestMethod("GET");
                    myURLConnection.setRequestProperty("X-NCU-API-TOKEN", getString(R.string.ncu_api_token));
                    myURLConnection.setUseCaches(false);
                    myURLConnection.setDoInput(true);
                    myURLConnection.connect();

                    int response = myURLConnection.getResponseCode();
                    BufferedReader r = new BufferedReader(new InputStreamReader(myURLConnection.getInputStream()));
                    StringBuilder total = new StringBuilder();
                    String line;
                    while ((line = r.readLine()) != null) {
                        total.append(line);
                    }
                    result = total.toString();

                    if (response == 200) {
                        JSONArray jsonArray = new JSONArray(result);

                        for (int i = 0; i < jsonArray.length(); i++) {

                            try {
                                JSONObject jsonData = jsonArray.getJSONObject(i);
                                String name = jsonData.getString("name");
                                String content = jsonData.getString("content");
                                String time = jsonData.getString("start");
                                String club = jsonData.getString("club");
                                String place = jsonData.getString("place");
                                String time_parsed = time;
                                Date date;
                                SimpleDateFormat simple = new java.text.SimpleDateFormat();
                                simple.applyPattern("yyyy-MM-dd HH:mm");
                                if(time != "null") {
                                    date = simple.parse(time);
                                    Calendar calendar = Calendar.getInstance();
                                    calendar.setTime(date);
                                    if (calendar.get(Calendar.HOUR_OF_DAY) == 0) {
                                        simple.applyPattern("yyyy-MM-dd");
                                        time_parsed = simple.format(date);
                                    }
                                }else{
                                    time = "無日期";
                                    time_parsed = "無日期";
                                }
                                ListData listData;
                                if (club != "null") {
                                    listData = new ListData(club + "-" + name, place, time, content, time_parsed);
                                } else {
                                    listData = new ListData(name, place, time, content, time_parsed);
                                }
                                itemsArray.add(listData);
                            }
                            catch (JSONException je) {
                                    Log.e("Debug", je.toString());
                            }
                        }
                    }


                } catch (Exception e) {
                    e.printStackTrace();
                    Log.e("Debug", e.toString());
                    serverStatus = false;
                }

            } else {
                Log.e("Debug", "No network");
                isNetwork = false;
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void arg0) {
            if (!isNetwork) {
                AlertDialog.Builder MyAlertDialog = new AlertDialog.Builder(getActivity());
                MyAlertDialog.setTitle("錯誤~");
                MyAlertDialog.setMessage("需要網際網路連線");
                MyAlertDialog.show();
            }
            if(!serverStatus){
                Toast.makeText(getActivity(), "無法與伺服器連線", Toast.LENGTH_SHORT).show();
            }
            listAdapter = new MyExpandableListItemAdapter(
                    getActivity(), itemsArray);
            alphaInAnimationAdapter = new AlphaInAnimationAdapter(listAdapter);
            alphaInAnimationAdapter.setAbsListView(m_ListView);
            assert alphaInAnimationAdapter.getViewAnimator() != null;
            alphaInAnimationAdapter.getViewAnimator().setInitialDelayMillis(INITIAL_DELAY_MILLIS);
            m_ListView.setAdapter(alphaInAnimationAdapter);

            if (progressDialog != null)
                progressDialog.dismiss();

        }
    }
}
